File: apps/native/modules/uoplan-engine/android/src/main/java/party/uoplan/engine/UoplanEngineModule.kt
package party.uoplan.engine

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/** Thrown when the native engine cannot be constructed or a generation call fails. */
internal class EngineException(message: String) : CodedException(message)

/**
 * Expo native module wrapping the Rust schedule-generation engine, linked from
 * `libuoplan_engine.so` (built per-ABI by `pnpm build:engine-native-ffi-android`).
 * The whole surface is bytes-in / bytes-out (protobuf `ByteArray` ↔ `Uint8Array`),
 * matching the JS `EngineBridge` contract; the same engine runs as WASM on the web
 * and through the C ABI on iOS, so generated timetables are identical.
 *
 * The `external fun`s map to the `Java_party_uoplan_engine_UoplanEngineModule_*`
 * JNI exports in `packages/engine/src/jni_android.rs`.
 */
class UoplanEngineModule : Module() {
  /** Serialises access to the (non-thread-safe) engine handle. */
  private val lock = Any()

  /** Opaque pointer to the Rust `EngineCore` (0 = none). */
  private var handle: Long = 0L

  /** The dataset (term) currently loaded, so repeated generations skip decode. */
  private var loadedKey: String? = null

  companion object {
    init {
      System.loadLibrary("uoplan_engine")
    }

    @JvmStatic external fun nativeNew(catalogue: ByteArray, schedules: ByteArray): Long

    @JvmStatic external fun nativeGenerate(handle: Long, request: ByteArray): ByteArray?

    @JvmStatic external fun nativeTimetableFixedSet(handle: Long, request: ByteArray): ByteArray?

    @JvmStatic external fun nativeFree(handle: Long)
  }

  override fun definition() = ModuleDefinition {
    Name("UoplanEngine")

    // Construct (or reuse) the engine for a dataset. Catalogue + schedules are the
    // encoded `Catalogue` / `SchedulesData` protobufs. No-op when already loaded.
    AsyncFunction("loadDataset") { key: String, catalogue: ByteArray, schedules: ByteArray ->
      synchronized(lock) {
        if (loadedKey == key && handle != 0L) {
          return@synchronized true
        }
        freeHandleLocked()
        val newHandle = nativeNew(catalogue, schedules)
        if (newHandle == 0L) {
          throw EngineException("failed to construct engine (dataset failed to decode)")
        }
        handle = newHandle
        loadedKey = key
        true
      }
    }

    // Generate a schedule for a serialized `GenerationRequest` → `GenerationResponse`.
    AsyncFunction("generate") { request: ByteArray ->
      runEngine(request) { h, req -> nativeGenerate(h, req) }
    }

    // Re-timetable a fixed set of courses for a serialized `TimetableRequest`.
    AsyncFunction("timetableFixedSet") { request: ByteArray ->
      runEngine(request) { h, req -> nativeTimetableFixedSet(h, req) }
    }

    OnDestroy {
      synchronized(lock) { freeHandleLocked() }
    }
  }

  /** Run a bytes-in/bytes-out engine call under the lock. */
  private fun runEngine(request: ByteArray, op: (Long, ByteArray) -> ByteArray?): ByteArray {
    synchronized(lock) {
      val h = handle
      if (h == 0L) {
        throw EngineException("engine not loaded; call loadDataset first")
      }
      return op(h, request) ?: throw EngineException("engine call failed")
    }
  }

  /** Release the current handle. Caller must hold `lock`. */
  private fun freeHandleLocked() {
    if (handle != 0L) {
      nativeFree(handle)
      handle = 0L
    }
    loadedKey = null
  }
}

File: apps/native/modules/uoplan-engine/ios/UoplanEngineModule.swift
import ExpoModulesCore
import UoplanEngineFFI

/// Thrown when the native engine cannot be constructed or a generation call fails.
internal final class EngineException: GenericException<String>, @unchecked Sendable {
  override var reason: String { param }
}

/// Expo native module wrapping the Rust schedule-generation engine (linked from
/// the `UoplanEngine.xcframework` static library). The whole surface is
/// bytes-in / bytes-out (protobuf `Uint8Array`), matching the JS `EngineBridge`
/// contract; the same engine runs as WASM on the web, so results are identical.
public final class UoplanEngineModule: Module {
  /// Serialises access to the (non-thread-safe) engine handle.
  private let lock = NSLock()
  /// Opaque pointer to the Rust `EngineCore`.
  private var handle: UnsafeMutableRawPointer?
  /// The dataset (term) currently loaded, so repeated generations skip the
  /// expensive catalogue decode.
  private var loadedKey: String?

  public func definition() -> ModuleDefinition {
    Name("UoplanEngine")

    // Construct (or reuse) the engine for a dataset. Catalogue + schedules are
    // the encoded `Catalogue` / `SchedulesData` protobufs. No-op when the same
    // dataset is already loaded.
    AsyncFunction("loadDataset") { (key: String, catalogue: Data, schedules: Data) -> Bool in
      self.lock.lock()
      defer { self.lock.unlock() }

      if self.loadedKey == key, self.handle != nil {
        return true
      }
      self.freeHandleLocked()

      let newHandle: UnsafeMutableRawPointer? = catalogue.withUnsafeBytes { catRaw in
        schedules.withUnsafeBytes { schedRaw in
          uoplan_engine_new(
            catRaw.bindMemory(to: UInt8.self).baseAddress,
            catalogue.count,
            schedRaw.bindMemory(to: UInt8.self).baseAddress,
            schedules.count
          )
        }
      }
      guard let newHandle else {
        throw EngineException("failed to construct engine (dataset failed to decode)")
      }
      self.handle = newHandle
      self.loadedKey = key
      return true
    }

    // Generate a schedule for a serialized `GenerationRequest`; returns a
    // serialized `GenerationResponse`.
    AsyncFunction("generate") { (request: Data) -> Data in
      try self.runEngine(request) { handle, ptr, len, outLen in
        uoplan_engine_generate(handle, ptr, len, outLen)
      }
    }

    // Re-timetable a fixed set of courses for a serialized `TimetableRequest`.
    AsyncFunction("timetableFixedSet") { (request: Data) -> Data in
      try self.runEngine(request) { handle, ptr, len, outLen in
        uoplan_engine_timetable_fixed_set(handle, ptr, len, outLen)
      }
    }

    OnDestroy {
      self.lock.lock()
      defer { self.lock.unlock() }
      self.freeHandleLocked()
    }
  }

  /// Run a bytes-in/bytes-out engine call under the lock, copying the result out
  /// of the Rust-owned buffer and freeing it.
  private func runEngine(
    _ request: Data,
    _ op: (UnsafeMutableRawPointer?, UnsafePointer<UInt8>?, Int, UnsafeMutablePointer<Int>?) -> UnsafeMutablePointer<UInt8>?
  ) throws -> Data {
    self.lock.lock()
    defer { self.lock.unlock() }

    guard let handle = self.handle else {
      throw EngineException("engine not loaded; call loadDataset first")
    }
    var outLen = 0
    let resultPtr: UnsafeMutablePointer<UInt8>? = request.withUnsafeBytes { reqRaw in
      op(handle, reqRaw.bindMemory(to: UInt8.self).baseAddress, request.count, &outLen)
    }
    guard let resultPtr else {
      throw EngineException("engine call failed")
    }
    let data = Data(bytes: resultPtr, count: outLen)
    uoplan_engine_free_buf(resultPtr, outLen)
    return data
  }

  /// Release the current handle. Caller must hold `lock`.
  private func freeHandleLocked() {
    if let handle = self.handle {
      uoplan_engine_free(handle)
      self.handle = nil
    }
    self.loadedKey = nil
  }
}
